libp2p/protocols/relay.nim

```diff
@@ -0,0 +1,471 @@
+    trace "unexcepted relay stop response", msgType = msgRcvFromRelay.msgType
+    await sendStatus(conn, RelayStatus.HopCantOpenDstStream)
+    raise newException(RelayError, "Hop can't open destination stream")
+
+  if msgRcvFromRelay.status.isNone or msgRcvFromRelay.status.get() != RelayStatus.Success:
+    trace "relay stop failure", status=msgRcvFromRelay.status
+    await sendStatus(conn, RelayStatus.HopCantOpenDstStream)
+    raise newException(RelayError, "Hop can't open destination stream")
+  result = conn
+
+#
+# Relay Transport
+#
+
+const
+  CircuitRelay = mapEq("p2p-circuit")
+  P2PPat = mapEq("p2p")
+
+type
+  RelayTransport* = ref object of Transport
+    relay: Relay
+
+method start*(self: RelayTransport, ma: MultiAddress) {.async.} =
+  if self.running:
+    trace "Relay transport already running"
+    return
+
+  await procCall Transport(self).start(ma)
+  trace "Starting Relay transport"
+
+method stop*(self: RelayTransport) {.async, gcsafe.} =
+  self.running = false
+  while not self.relay.queue.empty():
+    await self.relay.queue.popFirstNoWait().close()
+
+method accept*(self: RelayTransport): Future[Connection] {.async, gcsafe.} =
+  result = await self.relay.queue.popFirst()
+
+# TODO: dial / handles active relay:
+# /ipv4/127.0.0.1/tcp/5002/p2p/QmRelay/p2p-circuit/ipv4/127.0.0.1/tcp/5003/p2p/QmTarget
+# dstPeerId = QmTarget
+# dstAddrs = /ipv4/127.0.0.1/tcp/5003
+#
+# TODO: Must find a way to solve the handles problem, currently the dst switch is created
+# with two multiaddress ('/ip4/.../tcp/...' and '/ip4/.../tcp/.../p2p-circuit') in order to
+# works properly because with no circuit-relay address the transport won't be started.
+# It should work only with the first one.
+method dial*(self: RelayTransport, ma: MultiAddress): Future[Connection] {.async, gcsafe.} =
+  let
+    sma = toSeq(ma.items())
+    relayAddrs = sma[0..sma.len-4].mapIt(it.tryGet()).foldl(a & b)
+  var
+    relayPeerId: PeerId
+    dstPeerId: PeerId
+  if not relayPeerId.init(($(sma[^3].get())).split('/')[2]):
+    # error
+    discard
+  if not dstPeerId.init(($(sma[^1].get())).split('/')[2]):
+    # error
+    discard
+  trace "Dial", relayPeerId, relayAddrs, dstPeerId
+
+  let conn = await self.relay.switch.dial(relayPeerId, @[ relayAddrs ], RelayCodec)
+  result = await self.relay.dialPeer(conn, dstPeerId, @[])
+
+method dial*(
+  self: RelayTransport,
+  hostname: string,
+  address: MultiAddress): Future[Connection] {.async, gcsafe.} =
+  result = await self.dial(address)
+
+method handles*(self: RelayTransport, ma: MultiAddress): bool {.gcsafe} =
+  result=false
+  if ma.protocols.isOk:
+    let sma = toSeq(ma.items())
+    if sma.len >= 3:
+      result=(CircuitRelay.match(sma[^2].get()) and P2PPat.match(sma[^1].get())) or CircuitRelay.match(sma[^1].get())
+  trace "Handles return", ma, result
+
+proc new*(T: typedesc[RelayTransport], relay: Relay): T =
+  T(relay: relay, upgrader: relay.switch.transports[0].upgrader)
```

Comment:
There isn't any need to do this, you can safelly override `upgradeIncoming` and `upgradeOutgoing` to do nothing and it would work as expected.

Reply:
Edit: we actually need to upgrade, though adding support in `builders.nim` is a better solution

Reply:
Ah, you're right, we need to upgrade the connection and correct, we can just pass it as a parameter. 

I wonder if we can pass the `transports` array as well without having to expose it in the switch - we should be able to too.